File: python/src/application.py
import glob
import os
import re
from typing import Any


class Application:
    """Base class for the wiki-organising commands. Loads the wiki tree
    under base_path, groups it by Owner or Category in either English or
    Japanese, and exposes the resulting maps to its subclasses."""

    class NotImplementedError(Exception):
        pass

    NO_DECLARATION = {
        ('Owner', 'English'): 'Unowned',
        ('Owner', 'Japanese'): 'Owner記名なし',
        ('Category', 'English'): 'Uncategorised',
        ('Category', 'Japanese'): 'Category記載なし',
    }

    TARGET_REGEXP = {
        'Owner': re.compile(r'[Oo]wner:\s?'),
        'Category': re.compile(r'[Cc]ategory:\s?'),
    }

    _cached_wiki_maps: dict[str, list[str]]

    @classmethod
    def run(cls, base_path: str = os.path.join('..', '..'),
            group_by: str = 'Owner', language: str = 'English',
            home_overflow: str | bool = 'false') -> Any:
        instance = cls(
            base_path=base_path, group_by=group_by,
            language=language, home_overflow=home_overflow,
        )
        instance.validate()
        return instance._run()

    def __init__(self, base_path: str, group_by: str, language: str,
                 home_overflow: str | bool) -> None:
        self._base_path = base_path
        self._group_by = group_by
        self._language = language
        self._home_overflow = self._parse_home_overflow(home_overflow)
        self._path_to_home = os.path.join(base_path, 'Home.md')
        self._path_to_sidebar = os.path.join(base_path, '_Sidebar.md')
        self._path_to_github_wiki_organisers = sorted(glob.glob(
            os.path.join(base_path, 'github-wiki-organisers', '**', '*.md'),
            recursive=True))
        self._path_to_wikis_by_owner = sorted(glob.glob(
            os.path.join(base_path, 'wikis-by-owner', '*.md')))
        self._paths_to_wikis = sorted(glob.glob(
            os.path.join(base_path, '**', '*.md'), recursive=True))

    def validate(self) -> None:
        if self._group_by not in ('Owner', 'Category'):
            raise ValueError(f'Invalid group_by: `{self._group_by}`')
        if self._language not in ('English', 'Japanese'):
            raise ValueError(f'Invalid language: `{self._language}`')
        if self._home_overflow not in (True, False):
            raise ValueError(
                f'Invalid home_overflow: `{self._home_overflow}` '
                'must be boolean')

    # private

    def _run(self) -> Any:
        raise self.NotImplementedError(
            'This method must be implemented in each subclass.')

    @staticmethod
    def _parse_home_overflow(raw: str | bool) -> str | bool:
        match raw:
            case 'true':
                return True
            case 'false':
                return False
            case _:
                return raw

    def _target_paths(self) -> list[str]:
        excluded = {self._path_to_home, self._path_to_sidebar}
        excluded.update(self._path_to_github_wiki_organisers)
        excluded.update(self._path_to_wikis_by_owner)
        return [p for p in self._paths_to_wikis if p not in excluded]

    def _target_regexp(self) -> re.Pattern[str]:
        return self.TARGET_REGEXP.get(self._group_by, re.compile(''))

    def _no_declaration(self) -> str:
        return self.NO_DECLARATION.get(
            (self._group_by, self._language), '')

    def _wiki_maps_with_namespace(self) -> dict[str, list[str]]:
        if hasattr(self, '_cached_wiki_maps'):
            return self._cached_wiki_maps
        named: dict[str, list[str]] = {}
        uncategorised: dict[str, list[str]] = {}
        for target_path in self._target_paths():
            self._populate_namespace(target_path, named, uncategorised)
        result = dict(sorted(named.items()))
        result.update(uncategorised)
        self._cached_wiki_maps = result
        return result

    def _populate_namespace(self, target_path: str,
                            named: dict[str, list[str]],
                            uncategorised: dict[str, list[str]]) -> None:
        if not os.path.isfile(target_path):
            return
        with open(target_path) as f:
            wiki = os.path.basename(target_path)
            namespace = self._namespace_for(f)
        target = uncategorised if namespace == self._no_declaration() \
            else named
        target.setdefault(namespace, []).append(wiki)

    def _namespace_for(self, file: Any) -> str:
        declaration = file.readline()
        if not declaration or not self._target_regexp().match(declaration):
            return self._no_declaration()
        return re.sub(self._target_regexp(), '', declaration.rstrip('\n'))

    def _owned_wiki_maps(self) -> dict[str, list[str]]:
        return {
            namespace: wikis
            for namespace, wikis in self._wiki_maps_with_namespace().items()
            if '@' in namespace
        }

    def _plain_wiki_maps(self) -> dict[str, list[str]]:
        return {
            namespace: wikis
            for namespace, wikis in self._wiki_maps_with_namespace().items()
            if '@' not in namespace
        }

File: ruby/src/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

# Base class for the wiki-organising commands. Loads the wiki tree under
# base_path, groups it by Owner or Category in either English or Japanese, and
# exposes the resulting maps to its subclasses.
class Application
  class NotImplementedError < StandardError; end

  NO_DECLARATION = {
    %w[Owner English]     => 'Unowned',
    %w[Owner Japanese]    => 'Owner記名なし',
    %w[Category English]  => 'Uncategorised',
    %w[Category Japanese] => 'Category記載なし'
  }.freeze

  TARGET_REGEXP = {
    'Owner'    => /[Oo]wner:\s?/,
    'Category' => /[Cc]ategory:\s?/
  }.freeze

  # @rbs base_path: String
  # @rbs group_by: String
  # @rbs language: String
  # @rbs home_overflow: String
  # @rbs return: void
  def self.run(base_path: File.join('..', '..'), group_by: 'Owner', language: 'English', home_overflow: 'false')
    instance = new(base_path:, group_by:, language:, home_overflow:)
    instance.validate!
    instance.run
  end

  # @rbs base_path: String
  # @rbs group_by: String
  # @rbs language: String
  # @rbs home_overflow: String
  # @rbs return: void
  def initialize(base_path:, group_by:, language:, home_overflow:)
    @base_path     = base_path
    @group_by      = group_by
    @language      = language
    @home_overflow = parse_home_overflow(home_overflow)
    @path_to_home                   = File.join(base_path, 'Home.md')
    @path_to_sidebar                = File.join(base_path, '_Sidebar.md')
    @path_to_github_wiki_organisers = Dir[File.join(base_path, 'github-wiki-organisers', '**', '*.md')]
    @path_to_wikis_by_owner         = Dir[File.join(base_path, 'wikis-by-owner', '*.md')]
    @paths_to_wikis                 = Dir[File.join(base_path, '**', '*.md')]
  end

  # @rbs return: void
  def validate!
    raise ArgumentError, "Invalid group_by: `#{group_by}`" unless %w[Owner Category].include?(group_by)
    raise ArgumentError, "Invalid language: `#{language}`" unless %w[English Japanese].include?(language)

    return if [true, false].include?(home_overflow)

    raise ArgumentError, "Invalid home_overflow: `#{home_overflow}` must be boolean"
  end

  # @rbs return: void
  def run
    raise NotImplementedError, 'This method must be implemented in each subclass.'
  end

  private

  attr_reader :base_path,
              :group_by,
              :language,
              :home_overflow,
              :path_to_home,
              :path_to_sidebar,
              :path_to_github_wiki_organisers,
              :path_to_wikis_by_owner,
              :paths_to_wikis

  # @rbs raw: String
  # @rbs return: (bool | String)
  def parse_home_overflow(raw)
    case raw
    when 'true'  then true
    when 'false' then false
    else raw
    end
  end

  # @rbs return: Array[String]
  def target_paths
    @target_paths ||= paths_to_wikis.delete_if do |path_to_wiki|
      path_to_wiki == path_to_home ||
        path_to_wiki == path_to_sidebar ||
        path_to_github_wiki_organisers.include?(path_to_wiki) ||
        path_to_wikis_by_owner.include?(path_to_wiki)
    end
  end

  # @rbs return: Regexp
  def target_regexp
    @target_regexp ||= TARGET_REGEXP.fetch(group_by, //)
  end

  # @rbs return: String
  def no_declaration
    @no_declaration ||= NO_DECLARATION.fetch([group_by, language], '')
  end

  # @rbs array: Array[untyped]
  # @rbs hash: Hash[String, Array[untyped]]
  # @rbs return: Hash[String, Array[String]]
  def wiki_maps_with_namespace(array = [], hash = Hash.new { |h, namespace| h[namespace] = array.dup })
    uncategorised = hash

    @wiki_maps_with_namespace ||= target_paths.each.with_object(hash.dup) do |target_path, acc|
      populate_namespace(target_path, acc, uncategorised)
    end.sort.to_h.merge(uncategorised)
  end

  # @rbs target_path: String
  # @rbs hash: Hash[String, Array[String]]
  # @rbs uncategorised: Hash[String, Array[String]]
  # @rbs return: void
  def populate_namespace(target_path, hash, uncategorised)
    return unless File.exist?(target_path)

    File.open(target_path) do |file|
      wiki      = File.basename(file)
      namespace = namespace_for(file)

      target = namespace == no_declaration ? uncategorised : hash
      target[namespace] << wiki
    end
  end

  # @rbs file: File
  # @rbs return: String
  def namespace_for(file)
    declaration = file.readlines.first
    return no_declaration unless declaration&.start_with?(target_regexp)

    declaration.chomp.gsub(target_regexp, '')
  end

  # @rbs return: Hash[String, Array[String]]
  def owned_wiki_maps
    @owned_wiki_maps ||= wiki_maps_with_namespace.select { |namespace, _| namespace.include?('@') }
  end

  # @rbs return: Hash[String, Array[String]]
  def plain_wiki_maps
    @plain_wiki_maps ||= wiki_maps_with_namespace.reject { |namespace, _| namespace.include?('@') }
  end
end
